core/domain/classifier_services.py

```diff
@@ -0,0 +1,322 @@
+
+        label_set = set(
+            ['_default'] +
+            [label for labels in labels_list for label in labels])
+
+        self._label_count = len(label_set)
+        self._label_to_id = dict(zip(label_set, xrange(self._label_count)))
+
+        self._word_count = 0
+        self._word_to_id = {}
+
+        self._doc_count = len(docs)
+
+        self._b_dl = numpy.array(
+            map(self._get_label_vector, labels_list), dtype=int)
+        self._w_dc = [map(self._get_word_id, doc) for doc in docs]
+        self._l_dc = []
+        self._c_dl = numpy.zeros(
+            (self._doc_count, self._label_count), dtype=int)
+        self._c_lw = numpy.zeros(
+            (self._label_count, self._word_count), dtype=int)
+        self._c_l = numpy.zeros(self._label_count, dtype=int)
+
+        self._init_docs()
+        self._train_docs(iterations)
+
+    def add_examples(self, examples, iterations=5):
+        """Adds examples. Old examples are preserved."""
+        docs, labels_list = self._parse_examples(examples)
+
+        last_label_count = self._label_count
+        last_doc_count = self._doc_count
+        last_word_count = self._word_count
+
+        [map(self._get_label_id, labels) for labels in labels_list]
+        self._doc_count += len(docs)
+
+        if len(examples) > 0:
+            self._b_dl = numpy.concatenate(
+                (self._b_dl, numpy.zeros(
+                    (last_doc_count, self._label_count - last_label_count),
+                    dtype=int)), axis=1)
+            self._b_dl = numpy.concatenate(
+                (
+                    self._b_dl,
+                    [self._get_label_vector(labels) for labels in labels_list]
+                ), axis=0)
+            self._w_dc.extend([map(self._get_word_id, doc) for doc in docs])
+            self._c_dl = numpy.concatenate(
+                (self._c_dl, numpy.zeros(
+                    (last_doc_count, self._label_count - last_label_count),
+                    dtype=int)), axis=1)
+            self._c_dl = numpy.concatenate(
+                (self._c_dl, numpy.zeros(
+                    (self._doc_count - last_doc_count, self._label_count),
+                    dtype=int)), axis=0)
+            self._c_lw = numpy.concatenate(
+                (self._c_lw, numpy.zeros(
+                    (last_label_count, self._word_count - last_word_count),
+                    dtype=int)), axis=1)
+            self._c_lw = numpy.concatenate(
+                (self._c_lw, numpy.zeros(
+                    (self._label_count - last_label_count, self._word_count),
+                    dtype=int)), axis=0)
+            self._c_l = numpy.concatenate(
+                (self._c_l, numpy.zeros(
+                    self._label_count - last_label_count, dtype=int)))
+
+        for d in xrange(last_doc_count, self._doc_count):
+            self._init_docs([d])
+            self._train_docs(iterations, [d])
+
+        return xrange(last_doc_count, self._doc_count)
```

Comment:
From the method name, I wouldn't have expected this to return anything. Please add information to the docstring about what's being returned.